File: domain/file.go
package domain

import "time"

// File represents a file stored in Telegram and tracked in the virtual file system.
type File struct {
	ID                int        `db:"id" json:"id"`
	UserID            int        `db:"user_id" json:"user_id"`
	FolderID          *int       `db:"folder_id" json:"folder_id"`
	Name              string     `db:"name" json:"name"`
	Size              int64      `db:"size" json:"size"`
	MimeType          string     `db:"mime_type" json:"mime_type"`
	TelegramMessageID int        `db:"telegram_message_id" json:"telegram_message_id"`
	TelegramChatID    int64      `db:"telegram_chat_id" json:"telegram_chat_id"`
	TelegramFileID    string     `db:"telegram_file_id" json:"telegram_file_id"`
	IsStarred         bool       `db:"is_starred" json:"is_starred"`
	DeletedAt         *time.Time `db:"deleted_at" json:"deleted_at,omitempty"`
	CreatedAt         time.Time  `db:"created_at" json:"created_at"`
	UpdatedAt         time.Time  `db:"updated_at" json:"updated_at"`
}

// FileRepository defines the interface for file persistence operations.
type FileRepository interface {
	// Create inserts a new file record and returns the created file.
	Create(file *File) (*File, error)

	// GetByID retrieves a file by its ID.
	GetByID(id int) (*File, error)

	// ListByFolder lists non-deleted files for a user in a given folder (nil = root).
	ListByFolder(userID int, folderID *int) ([]File, error)

	// Update updates the file name and/or folder (rename/move).
	Update(file *File) error

	// SoftDelete sets deleted_at on a file.
	SoftDelete(id int) error

	// Restore clears deleted_at on a file.
	Restore(id int) error

	// ListTrashed returns all soft-deleted files for a user.
	ListTrashed(userID int) ([]File, error)

	// PermanentDelete removes a file record permanently from the database.
	PermanentDelete(id int) error

	// PermanentDeleteAllTrashed permanently deletes all trashed files for a user.
	PermanentDeleteAllTrashed(userID int) error

	// ListStarred lists all non-deleted starred files for a user.
	ListStarred(userID int) ([]File, error)
}

File: src/domain/file.ts
export interface File {
  id: number;
  user_id: number;
  folder_id: number | null;
  name: string;
  size: number;
  mime_type: string;
  telegram_message_id: number;
  telegram_chat_id: string; // BIGINT as string in JS
  telegram_file_id: string;
  is_starred: boolean;
  deleted_at?: Date | null;
  created_at: Date;
  updated_at: Date;
}

export interface FileRepository {
  create(file: {
    user_id: number;
    folder_id: number | null;
    name: string;
    size: number;
    mime_type: string;
    telegram_message_id: number;
    telegram_chat_id: string;
    telegram_file_id: string;
  }): Promise<File>;
  getById(id: number): Promise<File | null>;
  listByFolder(userId: number, folderId: number | null): Promise<File[]>;
  update(file: Pick<File, 'id' | 'name' | 'folder_id' | 'is_starred'>): Promise<void>;
  softDelete(id: number): Promise<void>;
  restore(id: number): Promise<void>;
  listTrashed(userId: number): Promise<File[]>;
  permanentDelete(id: number): Promise<void>;
  permanentDeleteAllTrashed(userId: number): Promise<void>;
  listStarred(userId: number): Promise<File[]>;
}
